node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -283,37 +306,61 @@ impl Initialized {
 		self.participation.process_active_leaves_update(ctx, &update).await?;
 
 		if let Some(new_leaf) = update.activated {
-			match self
-				.rolling_session_window
-				.cache_session_info_for_head(ctx.sender(), new_leaf.hash)
-				.await
-			{
-				Err(e) => {
-					gum::warn!(
+			let session_idx =
+				self.runtime_info.get_session_index_for_child(ctx.sender(), new_leaf.hash).await;
+
+			match session_idx {
+				Ok(session_idx)
+					if self.last_consecutive_cached_session.is_none() ||
+						session_idx >
+							self.last_consecutive_cached_session.expect(
+								"The first clause explicitly handles `None` case. qed.",
+							) =>
+				{
+					// There is a new session. Perform a dummy fetch to cache it.
+					let mut gap_in_cache = false;
+					for idx in self
+						.last_consecutive_cached_session
+						.unwrap_or(session_idx.saturating_sub(DISPUTE_WINDOW.get() - 2)) +
+						1..=session_idx
+					{
+						if let Err(err) = self
+							.runtime_info
+							.get_session_info_by_index(ctx.sender(), new_leaf.hash, idx)
+							.await
+						{
+							gum::debug!(
+								target: LOG_TARGET,
+								session_idx,
+								leaf_hash = ?new_leaf.hash,
+								?err,
+								"Error caching SessionInfo on ActiveLeaves update"
+							);
+							gap_in_cache = true;
+						}
+
+						if !gap_in_cache {
```

Comment:
What does `last_consecutive_cached_session` stand for exactly ?  For example if we have this this situation: [ fail ok, ok, ok, ok] ?

Reply:
We want the last `DISPUTE_WINDOW` sessions cached. IF we have to cache sessions 1..5 and for some reason 3 fails, on next ActiveLeaves update we want to retry fetching it and start caching from 3. So in this case `last_consecutive_cached_session` will be set to 2.
Regarding your question: `[ fail ok, ok, ok, ok]`. If the failed session is with index X, then `last_consecutive_cached_session` will be set to x-1.